gems/sorbet-runtime/lib/types/props/private/setter_factory.rb

```diff
@@ -78,33 +82,48 @@ def self.build_setter_proc(klass, prop, rules)
           non_nil_type: T::Types::Base,
           klass: T.all(Module, T::Props::ClassMethods),
           validate: T.nilable(ValidateProc),
+          check_level: Symbol,
         )
         .returns(SetterProc)
       end
-      private_class_method def self.nilable_proc(prop, accessor_key, non_nil_type, klass, validate)
+      private_class_method def self.nilable_proc(prop, accessor_key, non_nil_type, klass, validate, check_level)
         proc do |val|
           if val.nil?
             instance_variable_set(accessor_key, nil)
-          # this use of recursively_valid? is intentional: unlike for
-          # methods, we want to make sure data at the 'edge'
-          # (e.g. models that go into databases or structs serialized
-          # from disk) are correct, so we use more thorough runtime
-          # checks there
-          elsif non_nil_type.recursively_valid?(val)
-            validate&.call(prop, val)
-            instance_variable_set(accessor_key, val)
           else
-            T::Props::Private::SetterFactory.raise_pretty_error(
-              klass,
-              prop,
-              non_nil_type,
-              val,
-            )
+            if T::Props::Private::SetterFactory.should_check?(check_level)
```

Comment:
Have this if-statement and method call in the setter is likely to eat up a lot of potential gains (as well as having a cost even for folks not using `checked(:tests)`). 

Ideally we'd push the check to load time, ie outside this proc. Then you could just use `attr_writer` instead of using a proc at all. You'll have to worry about load order if you do that, though.

Whatever you do, I'd recommend adding/modifying some benchmarks (there's already a folder with a few).